react/AppIcon/index.jsx

```diff
@@ -17,15 +17,27 @@ const FETCHING = 'fetching'
 export class AppIcon extends Component {
   constructor(props, context) {
     super(props, context)
-    const { app, domain, secure } = props
-    const preloaded = getPreloaded(app, domain, secure)
-    this.state = {
-      error: null,
-      icon: preloaded,
-      status: preloaded ? DONE : FETCHING
-    }
     this.isUnmounting = false
     this.handleError = this.handleError.bind(this)
+
+    try {
+      const { app, client } = props
+      const cozyURL = new URL(client.getStackClient().uri)
+      this.domain = cozyURL.host
+      this.secure = cozyURL.protocol === 'https:'
```

Comment:
AMHA the component should not store this.domain and this.secure.

1. It makes the constructor a bit bloated
2. It makes the component deal with low level stuff when it is pretty high level

Passing the client directly to loadFn would be cleaner amha.

If the backward compatibility is blocking here I think a wrapper to transform getPreloaded and loadFn so that they accept a client would be better as it would at some point completely disappear.

```js
const passDomainSecure = fnTakingDomainAndSecure => {
  if (fnTakingDomainAndSecure.length === 2) {
    return fnTakingDomainAndSecure // Already accepting the app + client
  }
  return function (app, client) => {
      const cozyURL = new URL(client.getStackClient().uri)
      const domain = cozyURL.host
      const secure = cozyURL.protocol === 'https:'
      return fnTakingDomainAndSecure.apply(this, [app, domain, secure])
  }
}

const loadFn = passDomainSecure(this.props.loadFn)
const result = await loadFn(this.props.client)

```

Reply:
Yeah good point, i'll add a TODO for next time.